test/e2e/scalability/timer/timer.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package timer
+
+import (
+	"time"
+
+	"bytes"
+	"fmt"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/perftype"
+)
+
+var now = time.Now
+
+type phase struct {
+	sequenceNumber int
+	name           string
+	startTime      time.Time
+	endTime        time.Time
+}
+
+func (phase *phase) end() {
+	phase.endTime = now()
+}
+
+func (phase *phase) label() string {
+	return fmt.Sprintf("%03d-%s", phase.sequenceNumber, phase.name)
+}
+
+func (phase *phase) duration() time.Duration {
+	endTime := phase.endTime
+	if endTime.IsZero() {
+		endTime = now()
+	}
+	return endTime.Sub(phase.startTime)
+}
+
+type TestPhaseTimer struct {
+	current  *phase
```

Comment:
Seems like your timer model does not support having multiple current phases. This is a common use-case and I think we should support it. For e.g in load test we are doing 2 rounds of RC scaling. We may want to have individual times for them and also one 'combined' phase. In this case, we can probably just add the 2 phases to get the combined one. But that is 1) not convenient 2) doesn't cover non-composable phases.